doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,279 @@
+- `next_refresh` will usually be the write frontier of the MV's storage collection, unless:
+  - The write frontier is 0. This happens when the first refresh hasn't completed yet, in which case we need to figure out the first refresh: we'll call `RefreshSchedule::round_up_timestamp` on the MV's `as_of`.
+  - The write frontier is `[]`. This happens when the MV is after its last refresh, and will never be refreshed again. We could represent this simply with a `NULL`.
+- `last_completed_refresh` is roughly `RefreshSchedule::round_down_timestamp(next_refresh - 1)`, plus special handling for the `[]` write frontier case. (`round_down_timestamp` will be a new function, which will be the same as `round_up_timestamp`, but backwards.)
+
+For seeing whether the last refresh is being late, the user can run `EXPLAIN TIMESTAMP` in `STRICT SERIALIZABLE` mode, and look at can respond immediately. If it's false, then the last refresh's completion is overdue. Another way to get the same information would be to check if `mz_materialized_view_refreshes.next_refresh < now()`.
+
+For showing cluster schedules, I'm thinking of doing a similar thing as in `mz_materialized_view_refresh_options`, because eventually the `SCHEDULE =` option will allow for specifying multiple schedules. So I'd create a new table in `mz_internal` called `mz_cluster_schedules`, and it would have a row for each non-trivial (non-`MANUAL`) schedule option of each cluster. This would currently be only `SCHEDULE = ON REFRESH`. Columns would be `(cluster_id text, schedule_type text, refresh_rehydration_time_estimate interval)`. In the `schedule_type`, we would currently always have "ON REFRESH". (Eventually, we'll probably also want a `next_scheduled_turn_on`, but this doesn't seem so urgent. It will get more important when we'll be choosing the warmup time automatically.)
+
+For seeing whether a cluster is currently turned on, the user can simply look at `mz_cluster_replicas`, because we currently turn clusters On/Off by just creating/dropping replicas. We might also add a builtin view for showing this information in a more focused way.
+
+For the automatic cluster scheduling history, the user can look at `mz_audit_events`. This has a `details` column, which is a JSON blob, where I'm planning to add the `reason` for turning on a cluster, i.e., which materialized views were in need of a refresh. (See Nikhil's comment [here](https://github.com/MaterializeInc/materialize/pull/26401#pullrequestreview-1981986544).) There is also the `mz_cluster_replica_history` view, which takes its info from `mz_audit_events`, and presents the info in a nicer form. I could add a new reason column to this view.
```

Comment:
Would love to hear a bit more about how you're planning to represent `reason`. It might make sense to introduce a bit of substructure, and have the `reason` field be an object indicating all policies that said yes.

Reply:
Done. Added the following:
"Also note that the `reason` could also be prepared to show reasons from other policies: it could itself be a collection of key-value pairs, where the keys are policy names (e.g., refresh), and the values have policy-specific structures. For refresh, it could be a list of the materialized view IDs that made us turn the cluster on."

Reply:
Sorry, I have to hit "resolve" on this thread now, because GitHub got into that weird buggy state where it snatches back the focus to this text field every few seconds... (I tried refreshing the window, as well as doing a resolve and then unresolve, but it's still happening.)

Reply:
Ok, writing the above comment worked around the github bug. So then I'll let you hit resolve after checking my update.